mcp/streamable.go

```diff
@@ -800,3 +869,86 @@ func (s *streamableClientConn) Close() error {
 	})
 	return s.closeErr
 }
+
+// establishSSE creates and manages the persistent SSE listening stream.
+// It is used for both the initial connection and all subsequent reconnect attempts,
+// using the Last-Event-ID header to resume a broken stream where it left off.
+// On a successful response, it delegates to handleSSE to process events;
+// on failure, it triggers the client's reconnect logic.
+func (s *streamableClientConn) establishSSE(opts *startSSEState) {
+	select {
+	case <-s.done:
+		return
+	default:
+	}
+
+	req, err := http.NewRequestWithContext(s.ctx, http.MethodGet, s.url, nil)
+	if err != nil {
+		return
+	}
+	s.mu.Lock()
+	if s._sessionID != "" {
+		req.Header.Set("Mcp-Session-Id", s._sessionID)
+	}
+	s.mu.Unlock()
+	if opts.lastEventID != "" {
+		req.Header.Set("Last-Event-ID", opts.lastEventID)
+	}
+	req.Header.Set("Accept", "text/event-stream")
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		// On connection error, schedule a retry.
+		s.scheduleReconnect(opts)
+		return
+	}
+
+	// Per the spec, a 405 response means the server doesn't support SSE streams at this endpoint.
+	if resp.StatusCode == http.StatusMethodNotAllowed {
+		resp.Body.Close()
+		return
+	}
+
+	if !strings.Contains(resp.Header.Get("Content-Type"), "text/event-stream") {
+		resp.Body.Close()
+		return
+	}
+
+	s.handleSSE(resp, opts)
+}
+
+// scheduleReconnect schedules the next SSE reconnect attempt after a delay.
+func (s *streamableClientConn) scheduleReconnect(opts *startSSEState) {
+	reconnOpts := s.ReconnectOptions
+	if opts.attempt >= reconnOpts.MaxRetries {
+		s.mu.Lock()
+		s.err = fmt.Errorf("connection failed after %d attempts", reconnOpts.MaxRetries)
+		s.mu.Unlock()
+		s.Close() // Close the connection to unblock any readers.
+		return
+	}
+
+	delay := calculateReconnectDelay(reconnOpts, opts.attempt)
+
+	select {
+	case <-s.done:
+		return
+	case <-time.After(delay):
+		opts.attempt++
+		s.establishSSE(opts)
```

Comment:
This is a recursive call to establishSSE, so in theory you could blow the stack.
Not sure what the right fix is.